client/cluster_error.go

```diff
@@ -31,3 +36,39 @@ func (ce *ClusterError) Detail() string {
 	}
 	return s
 }
+
+// IsBadConnection returns true if all the Errors are related to failure
+// to connect to servers.
+// TODO: find a more reliable way to detect this
+func (ce *ClusterError) IsBadConnection() bool {
+	for _, err := range ce.Errors {
+		switch typedErr := (err).(type) {
+		case *net.OpError:
+			switch opErr := (typedErr.Err).(type) {
+			case *net.AddrError:
+				// triggered by out of range port
+				continue
+			case *net.DNSError:
+				// couldn't resolve name
+				continue
+			case *os.SyscallError:
+				if strings.Contains(opErr.Error(), "connection refused") {
+					// port not open
+					continue
+				}
+				return false
+			default:
```

Comment:
I am not sure if timeout should also count as a connection error. we can try net.OpError.Timeout()?